src/Planner/PlannerJoins.cpp

```diff
@@ -561,7 +567,7 @@ JoinClausesAndActions buildJoinClausesAndActions(
                 if (!left_key_node->result_type->equals(*common_type))
                     left_key_node = &left_join_actions->addCast(*left_key_node, common_type, {});
 
-                if (!right_key_node->result_type->equals(*common_type))
+                if (!is_join_with_special_storage && !right_key_node->result_type->equals(*common_type))
```

Comment:
Is it okay not to do type conversion? Maybe it's better to add an alias after adding a cast.

Reply:
If types are `Nullable(T)` and `T` it's alright, HashJoin handles it. For the case for example `UInt64` and `UInt32` HashJoin will throw exception about type mismatch, and it should be expected, because StorageJoin has nothing to do with it and should not rebuild prepared hash-table (alright, we might always covert left table column to right table type handling values that cannot be converting e.g. with accurateCastOrNull, but it's more generic solution, that is not implemented yet).